File: shared/src/main/java/de/dreier/mytargets/shared/targets/scoringstyle/ScoringStyle.java
/*
 * Copyright (C) 2017 Florian Dreier
 *
 * This file is part of MyTargets.
 *
 * MyTargets is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 2
 * as published by the Free Software Foundation.
 *
 * MyTargets is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 */

package de.dreier.mytargets.shared.targets.scoringstyle;

import android.support.annotation.NonNull;
import android.support.annotation.StringRes;

import com.annimon.stream.Stream;

import de.dreier.mytargets.shared.models.Score;
import de.dreier.mytargets.shared.models.db.End;
import de.dreier.mytargets.shared.models.db.Shot;

import static de.dreier.mytargets.shared.SharedApplicationInstance.get;

public class ScoringStyle {

    public static final String MISS_SYMBOL = "M";
    private static final String X_SYMBOL = "X";
    private final String title;
    private final boolean showAsX;
    protected final int[][] points;
    private int maxScorePerShot;

    ScoringStyle(boolean showAsX, int[][] points) {
        this(null, showAsX, points);
    }

    public ScoringStyle(@StringRes int title, boolean showAsX, int... points) {
        this(get(title), showAsX, new int[][]{points});
    }

    private ScoringStyle(String title, boolean showAsX, int[][] points) {
        this.showAsX = showAsX;
        this.points = points;
        getMaxPoints();
        if (title == null) {
            this.title = getDescriptionString();
        } else {
            this.title = title;
        }
    }

    public ScoringStyle(boolean showAsX, int... points) {
        this(showAsX, new int[][]{points});
    }

    private int getMaxPoints() {
        maxScorePerShot = 0;
        for (int[] arrowPoints : points) {
            for (int point : arrowPoints) {
                if (point > maxScorePerShot) {
                    maxScorePerShot = point;
                }
            }
        }
        return maxScorePerShot;
    }

    @NonNull
    private String getDescriptionString() {
        String style = "";
        for (int i = 0; i < points[0].length; i++) {
            if (i + 1 < points[0].length && points[0][i] <= points[0][i + 1] && !(i == 0 && showAsX)) {
                continue;
            }
            if (!style.isEmpty()) {
                style += ", ";
            }
            style += zoneToString(i, 0);
            for (int a = 1; a < points.length; a++) {
                style += "/" + zoneToString(i, a);
            }
        }
        return style;
    }

    @Override
    public String toString() {
        return title;
    }

    public String zoneToString(int zone, int arrow) {
        if (isOutOfRange(zone)) {
            return MISS_SYMBOL;
        } else if (zone == 0 && showAsX) {
            return X_SYMBOL;
        } else {
            int value = getScoreByScoringRing(zone, arrow);
            if (value == 0) {
                return MISS_SYMBOL;
            }
            return String.valueOf(value);
        }
    }

    public int getScoreByScoringRing(int zone, int arrow) {
        if (isOutOfRange(zone)) {
            return 0;
        }
        return getPoints(zone, arrow);
    }

    protected int getPoints(int zone, int arrow) {
        return points[0][zone];
    }

    private boolean isOutOfRange(int zone) {
        return zone < 0 || zone >= points[0].length;
    }

    public Score getReachedScore(Shot shot) {
        Score score = new Score();
        score.reachedScore = getScoreByScoringRing(shot.scoringRing, shot.index);
        score.totalScore = maxScorePerShot;
        return score;
    }

    public Score getReachedScore(End end) {
        return Stream.of(end.getShots())
                .map(this::getReachedScore)
                .reduce(new Score(), Score::add);
    }
}

File: shared/src/main/java/de/dreier/mytargets/shared/targets/scoringstyle/ScoringStyle.kt
/*
 * Copyright (C) 2018 Florian Dreier
 *
 * This file is part of MyTargets.
 *
 * MyTargets is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 2
 * as published by the Free Software Foundation.
 *
 * MyTargets is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 */

package de.dreier.mytargets.shared.targets.scoringstyle

import android.util.Log
import androidx.annotation.StringRes
import de.dreier.mytargets.shared.SharedApplicationInstance
import de.dreier.mytargets.shared.models.Score
import de.dreier.mytargets.shared.models.db.Shot
import de.dreier.mytargets.shared.models.sum

open class ScoringStyle private constructor(
        val title: String?,
        private val showAsX: Boolean,
        private val pointMiss: Int,
        protected val points: Array<IntArray>
) {
    private val maxScorePerShot: List<Int> by lazy { points.map { it.max() ?: 0 } }

    internal constructor(showAsX: Boolean, points: Array<IntArray>) : this(null, showAsX, 0, points)

    constructor(showAsX: Boolean, points: IntArray) : this(showAsX, arrayOf<IntArray>(points))

    constructor(showAsX: Boolean, pointMiss: Int, points: IntArray) : this(null, showAsX, pointMiss, arrayOf<IntArray>(points))

    constructor(@StringRes title: Int, showAsX: Boolean, points: IntArray) : this(SharedApplicationInstance.getStr(title), showAsX, 0, arrayOf<IntArray>(points))

    private val descriptionString: String
        get() {
            var style = ""
            for (i in 0 until points[0].size) {
                if (i + 1 < points[0].size && points[0][i] <= points[0][i + 1] && !(i == 0 && showAsX)) {
                    continue
                }
                if (!style.isEmpty()) {
                    style += ", "
                }
                style += zoneToString(i, 0)
                for (a in 1 until points.size) {
                    style += "/" + zoneToString(i, a)
                }
            }
            return style
        }

    override fun toString() = title ?: descriptionString

    fun zoneToString(zone: Int, arrow: Int): String {
        return if (isOutOfRange(zone)) {
            MISS_SYMBOL
        } else if (zone == 0 && showAsX) {
            X_SYMBOL
        } else {
            val value = getPointsByScoringRing(zone, arrow)
            if (value == 0) {
                MISS_SYMBOL
            } else value.toString()
        }
    }

    fun getPointsByScoringRing(zone: Int, arrow: Int): Int {
        return if (isOutOfRange(zone)) pointMiss else getPoints(zone, arrow)
    }

    protected open fun getPoints(zone: Int, arrow: Int): Int {
        return points[0][zone]
    }

    private fun isOutOfRange(zone: Int): Boolean {
        return zone < 0 || zone >= points[0].size
    }

    fun getReachedScore(shot: Shot): Score {
        val maxScore = getMaxPointsPerShot(shot.index)
        if (shot.scoringRing == Shot.NOTHING_SELECTED) {
            return Score(maxScore)
        }
        val reachedScore = getPointsByScoringRing(shot.scoringRing, shot.index)
        return Score(reachedScore, maxScore)
    }

    open fun getMaxPointsPerShot(index: Int): Int {
        val i = if (index < maxScorePerShot.size) index else maxScorePerShot.size - 1
        return maxScorePerShot[i]
    }

    open fun getReachedScore(shots: List<Shot>): Score {
        return shots.map { getReachedScore(it) }.sum()
    }

    companion object {
        const val MISS_SYMBOL = "M"
        private const val X_SYMBOL = "X"
    }
}
